clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -103,90 +105,113 @@ class GarbageCollector(val rangeGetter: RangeGetter) extends Serializable {
-
     // Filter out addresses outside the storage namespace.
     // Returned addresses are either relative ones, or absolute ones that
     // are under the storage namespace.
     def getDeletableAddresses(rangeID: String): Iterator[String] = {
       println(s"getAddressesToDelete: get addresses for range $rangeID")
-      rangeGetter
+      val addresses = rangeGetter
         .getRangeEntries(rangeID, repo)
         .filter(e => e.addressType.isRelative || e.address.startsWith(storageNS))
         .map(e =>
           if (e.addressType.isRelative) e.address
           else StringUtils.removeStart(e.address, storageNS)
         )
+      addresses
     }
 
-    val expiredAddresses = cleanExpiredRangeIDs
-      .repartition(numAddressPartitions)
-      .flatMap(getDeletableAddresses)
+    def getDeletableAddressesUnion(it: Iterator[(String, Boolean)]): Iterator[(String, Boolean)] = {
+      val all = collection.mutable.Set[(String, Boolean)]()
+      it.foreach({
+        case (rangeID, expire) => {
+          all ++= getDeletableAddresses(rangeID).map((x) => (x, expire))
+        }
+      })
+      all.iterator
+    }
 
-    val keepAddresses =
-      keepRangeIDs
-        .repartition(numAddressPartitions)
-        .flatMap(getDeletableAddresses)
+    def readRanges(ranges: Dataset[(String, Boolean)], name: String): RDD[(String, Boolean)] = {
+      val approxNumRows = ranges.cache().rdd.countApprox(10000 /* msecs */, 0.9)
+      val numPartitions = (try {
+        approxNumRows.getFinalValue.high
+      } catch {
+        case e: Exception => {
+          println(s"readRanges: Approximate count for $name failed, count precisely ($e)")
+          ranges.count
+        }
+      }) / approxNumRangesToSpreadPerPartition
+      println(s"readRanges: use $numPartitions partitions for $name")
+
+      val all = ranges
+        .repartition(Math.max(1, Math.ceil(numPartitions)).toInt)
+        .rdd
+        .mapPartitionsWithIndex((index: Int, it: Iterator[(String, Boolean)]) => {
+          println(s"get addresses partition ${index}")
+          getDeletableAddressesUnion(it)
+        })
+      if (sampleFraction == 1) all
+      else {
+        println(s"readRanges: downsample to ${sampleFraction}")
+        all.sample(true, sampleFraction)
+      }
+    }
+
+    // Expire an object _only_ if it is always expired!
+    val shouldExpire = (a: Boolean, b: Boolean) => a & b
+
+    val dedupedRangeIDs = rangeIDs.rdd
+      .aggregateByKey(0, numRangePartitions)(
+        (u: Int, expire: Boolean) => u | (if (expire) 1 else 2),
+        (a, b: Int) => a | b
+      )
+      .map({ case (r, expireInt) =>
+        // Ranges appearing only as expire should expire.  Ranges appearing
+        // only as keep should be kept.  Ranges appearing as both expire and
+        // keep should still be kept - to protect their contents from
+        // expired ranges that intersect.
+        (r, expireInt == 1)
```

Comment:
```suggestion
      .aggregateByKey(true, numRangePartitions)(
        (expireAcc: Boolean, expire: Boolean) => expireAcc && expire,
        (expireAcc1, expireAcc2: Int) => expireAcc1 && expireAcc2
      )
      .collect()
```
?

I probably don't understand something...